genie-common-internal/src/main/java/com/netflix/genie/common/internal/util/DynamicPropertiesMapCache.java

```diff
@@ -0,0 +1,142 @@
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.common.internal.util;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Sets;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.core.env.EnumerablePropertySource;
+import org.springframework.core.env.Environment;
+import org.springframework.core.env.PropertySource;
+import org.springframework.core.env.PropertySources;
+import org.springframework.core.env.StandardEnvironment;
+
+import java.time.Duration;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+/**
+ * Utility class that produces a map of properties and their values if they match a given pattern.
+ * The map is cached for a fixed amount of time to avoid re-computing for each call.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class DynamicPropertiesMapCache {
+
+    private static final String CACHE_KEY = "properties-map";
+    private final Environment environment;
+    private final Pattern propertiesNamesPattern;
+    private final LoadingCache<String, Map<String, String>> cache;
+
+    /**
+     * Constructor.
+     *
+     * @param refreshInterval        the interval after which the map is re-calculated
+     * @param environment            the environment
+     * @param propertiesNamesPattern the pattern that properties need to match to get included in the map
+     */
+    public DynamicPropertiesMapCache(
+        final @NonNull Duration refreshInterval,
+        final Environment environment,
+        final Pattern propertiesNamesPattern
+    ) {
+        this.environment = environment;
+        this.propertiesNamesPattern = propertiesNamesPattern;
+        this.cache = Caffeine
+            .newBuilder()
+            .refreshAfterWrite(refreshInterval)
+            .initialCapacity(1)
+            .build(this::loadProperties);
+    }
+
+    /**
+     * Get the current map of properties and their values.
+     *
+     * @return an immutable map of property name and property value, for any property that matches the pattern that has
+     * a non-blank value
+     */
+    public Map<String, String> get() {
+        return this.cache.get(CACHE_KEY);
+    }
+
+    private Map<String, String> loadProperties(@NonNull final String propertiesKey) {
+        // There's only 1 item in this cache, so key is redundant
+        return DynamicPropertiesMapCache.createDynamicPropertiesMap(
```

Comment:
why the extra redirect to a static method?

Reply:
This is a utility class that takes snapshots of properties matching a given regex.
And also provides a utility object that does the same but caches.

Maybe i should disentangle the two.